src/SmoFacade/Server.cs

```diff
@@ -1,209 +1,223 @@
+    ///   A template string for backup location:
+    ///   {0} databaseName
+    ///   {1} time
+    /// </param>
+    public void LogBackup(string databaseName, string backupPathTemplate)
+    {
+      backupPathTemplate = backupPathTemplate ?? DefaultBackupPathTemplate(".trn");
+      Backup(databaseName, backupPathTemplate, BackupActionType.Log, BackupTruncateLogType.Truncate);
+    }
+
+    private void Backup(string databaseName, string backupPathTemplate, BackupActionType backupActionType,
+      BackupTruncateLogType truncateType)
+    {
+      var backup = new Backup { Action = backupActionType, Database = databaseName, LogTruncation = truncateType };
+      var bdi = new BackupDeviceItem(string.Format(backupPathTemplate,
+                                                   databaseName,
+                                                   DateTime.Now.ToString("yyyy_MM_dd_hhmmss_fff")),
+                                     DeviceType.File);
+
+      backup.Devices.Add(bdi);
+      backup.SqlBackup(_server);
+    }
+
+    /// <summary>
+    ///   Generate a full backup, not truncating the transaction log, and storing it in the default backup destination.
+    ///   TODO: we should support a backup path somehow with configuration
+    /// </summary>
+    /// <param name="databaseName">name of the database to backup</param>
+    /// <param name="backupPathTemplate">
+    ///   A template string for backup location:
+    ///   {0} databaseName
+    ///   {1} time
+    /// </param>
+    public void FullBackup(string databaseName, string backupPathTemplate)
+    {
+      backupPathTemplate = backupPathTemplate ?? DefaultBackupPathTemplate(".bak");
+      Backup(databaseName, backupPathTemplate, BackupActionType.Database, BackupTruncateLogType.NoTruncate);
+    }
+
+    /// <summary>
+    ///   Builds a backup path template from the server's default backup directory.
+    ///   If this path is not valid on the destination instance the restore process will fail.
+    /// </summary>
+    private string DefaultBackupPathTemplate(string extension)
+    {
+      return _server.BackupDirectory + "\\{0}_backup_{1}" + extension;
+    }
+
+    internal Microsoft.SqlServer.Management.Smo.Login ConstructLogin(LoginProperties loginProperties)
+    {
+      var login = new Microsoft.SqlServer.Management.Smo.Login(_server, loginProperties.Name) {
+        LoginType = loginProperties.LoginType, Sid = loginProperties.Sid,
+        DefaultDatabase = loginProperties.DefaultDatabase
+      };
+
+      if(loginProperties.LoginType == LoginType.SqlLogin) {
+        if(loginProperties.PasswordHash != null)
+          login.Create(loginProperties.PasswordHash, LoginCreateOptions.IsHashed);
+        else if(loginProperties.Password != null)
+          login.Create(loginProperties.Password);
+        else
+          throw new ArgumentException("Password or hash was not supplied for sql login.");
+      }
+      else {
+        login.Create();
+      }

```

Comment:
While creating the login object requires the _server the rest of this is logic pertaining to the login, so I liked it being in the login class.

Reply:
The point of this was to get rid of the indirect functions returning new objects via `ConstructLogin`.
Agreed. Somewhere we'll have to get the private `_server` member. So that'd have to be exposed to put this in the Login class.